.github/workflows/macos_tests.yaml

```diff
@@ -0,0 +1,126 @@
+name: MacOS Tests
+on:
+  push:
+    branches:
+      - main
+      - 'releases/**'
+  pull_request:
+    branches:
+      - main
+      - 'releases/**'
+
+# Restrict tests to the most recent commit.
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  build_test_app:
+    name: Build Test Client
+    runs-on: macos-latest
+    outputs:
+      matrix: ${{ steps.testGen.outputs.tests }}
+
+    steps:
+      - name: Clone repository
+        uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+
+      - name: Install dependencies
```

Comment:
Can we possibly move all dependencies under one singular job? Currently "build_test_app" and "functionaltests" jobs both install a number of dependencies, but these packages can be installed once and both "build_test_app" and "functionaltests" jobs can rely on it. Same comment for the Linux logic too.

Something like:

```c++
jobs:
  install-dependencies:
    name: Install Dependencies
    runs-on: macos-latest
    steps:
      - name: Checkout repository
        uses: actions/checkout@v3
        with:
          submodules: 'recursive'
      - name: Install Python and Homebrew
        run: |
          brew install python
          brew install ninja
      - name: Install Python packages
        run: |
          pip3 install -r requirements.txt
```

Reply:
Not really, because the building of the client and the running of the test happen on separate workers and they both need to have dependencies installed. And worse yet, the dependencies needed by both of these tasks are slightly different; `build_test_app` needs the build dependencies while `functionaltests` needs the runtime dependencies.

We use multiple workers so that the tests can run in parallel, and it's also how Github's matrix feature works under the hood.

Reply:
That makes sense, thanks for the additional context!